### PR TITLE
Users topic partitioning

### PR DESCRIPTION
**Pull request by gandreadis**
Right now, the Kubernetes logs show issues with Kafka brokers, so not sure whether it works.

### CODE REVIEW

**Comment by gandreadis** [comment]
@qurben I think I've got a basic partition setup working for the stock topics, could you try to reproduce it on your k8s setup as well?

**Comment by qurben** [comment]
The get on stock is broken here, because creating off-loads to a specific instance, where retrieving tries to load from the instance the request was done on. This fails.

**Comment by qurben** [comment]
It seems that the requests always come to the first instance.

**Comment by qurben** [comment]
When you create different request topics, for each service one. It might be possible to also partition that topic. When the response topic is not partitioned all events will find their way back to the rest service. When we use the requestid as key on the request/response topics it should be possible to make sure that the response events get assigned to the right topic.

**Comment by qurben** [comment]
Another problem which is still present in this branch is that order checkout is done for an entire order. This is not possible when orders contain stock items which are on different instances. The order service should emit events for each stock item such that every event gets assigned to the right partition. The order service should again aggregate if subtracting was succesful for all stock items.

**Comment by gandreadis** [comment]
Thanks for the in-depth analysis! It does seem like this is becoming more complex than expected. The problem with partitioning the REST requests, as well, is that the POST requests might end up on different stock-service-partitions (since there is no stock ID yet) than the subsequent GET requests for that stock item...

**Comment by qurben** [comment]
Generating ids in the rest service might be a solution for that.

**Comment by gandreadis** [comment]
On the practical side, yes. On the clean software engineering side, no, since that should be the responsibility of the stock service, which should also be able to run independently. But I'd say let's still do it, since this is an experimental project, anyway.

**Comment by qurben** [comment]
The payment events in PaymentConsumer in users are probably not partitioned right.

**Comment by qurben** [comment]
`paymentCreated` should probably emit an event with user id as key which contains information for the users service to subtract credit.

**Comment by qurben** [comment]
Or maybe the payment service should already make sure that it sends an event on a users topic.

**Comment by gandreadis** [comment]
What do you mean exactly with your second proposal?

**Comment by gandreadis** [comment]
We could just give all payment created events the same ID, so that they all end up in the 0-partition...

**Comment by qurben** [comment]
In paymentconsumer the users service listens for payments, for these payments you cannot be sure that the user exists in that instance.

**Comment by qurben** [comment]
When the payments service creates events on a users topic it can be a bit faster because the users service does not need to send another event to get to the right users service

**Comment by gandreadis** [comment]
Very true 👍

**Comment by gandreadis** [comment]
I've partitioned the payment create topic on user ID, now. This should in theory solve the problem: the user service only gets payments relevant to its partitioned user, and the payment service consumer gets all event since it isn't configured for partitioning

**Comment by qurben** [comment]
e2e gets stuck on checkout

File: service/payments/src/main/java/nl/tudelft/wdm/group1/payments/events/Producer.java [hunk 1]
```diff
@@ -21,7 +21,7 @@
 
     public void emitPaymentCreated(Payment payment) {
         logger.info("Producing [{}] -> {}", PaymentsTopics.PAYMENT_CREATED, payment);
-        this.kafkaTemplate.send(PaymentsTopics.PAYMENT_CREATED, payment);
+        this.kafkaTemplate.send(PaymentsTopics.PAYMENT_CREATED, payment.getUserId().toString(), payment);
```
Comment (qurben) [inline service/payments/src/main/java/nl/tudelft/wdm/group1/payments/events/Producer.java]: This topic is not partitioned right? Only the topics which are in `UsersTopics` are.

Comment (gandreadis) [reply service/payments/src/main/java/nl/tudelft/wdm/group1/payments/events/Producer.java]: ~Ahh op die fiets~ Good catch, I'll add a Bean config

Comment (gandreadis) [reply service/payments/src/main/java/nl/tudelft/wdm/group1/payments/events/Producer.java]: (apologies for not catching this myself, my local Kubernetes configuration has acted up lately and I have tried but failed repeatedly to test this myself)

**Comment by qurben** [comment]
The message does not seem to end up in the right users instance. A ResourceNotFoundException is thrown.

**Comment by qurben** [comment]
The payments get received on the wrong topic in the users service.

**Pull request merged**